primitives/sr-primitives/src/lib.rs

```diff
@@ -451,6 +436,27 @@ impl From<&'static str> for DispatchError {
 	}
 }
 
+/// This type specifies the outcome of dispatching a call to a module.
+///
+/// In case of failure an error specific to the module is returned.
+pub type DispatchOutcome = Result<(), DispatchError>;
+
+/// The outcome of inclusion of an extrinsic into a block.
+///
+/// This type is typically used in the context of `BlockBuilder` to signal that the extrinsic
+/// in question cannot be included. It is fair to say that a valid block doesn't contain any
+/// extrinsic that would have had a negative inclusion outcome. On successful inclusion this type
+/// supplies the result of the extrinsic dispatch.
+///
+/// Examples of reasons preventing inclusion in a block:
+/// - More block weight is required to process the extrinsic than is left in the block being built.
+///   This doesn't neccessarily mean that the extrinsic is invalid, since it can still be
+///   included in the next block if it has enough spare weight available.
+/// - The sender doesn't have enough funds to pay the transaction inclusion fee. Including such
+///   a transaction in the block doesn't make sense.
+/// - The extrinsic supplied a bad signature. This transaction won't become valid ever.
```

Comment:
I would make it very clear that a dispatch has two phases 1- *pre-dispatch stuff* and 2- *the dispatch itself*. 

A signature check, and any other code that we might put in `Executive::apply_extrinsic_xxx` is in the first group (`.check`). For instance a signature check gives you `InclusionError`

A weight check, nonce and other things that happen either in signedExtensions or dispatch code itself are in the second group. I think a wrong nonce gives you `Ok(DispatchError)` (while an all okay dispatch gives `Ok(())`). 

So I would make this distinction very clear in the docs. 

(take the above with a grain of slat and double check, I didn't look super deep into the code.)

Reply:
Based on this, without any explanation, I find it confusing that signature check and weight check are enumerated next to each other, as it seems that they are natively different types of checks.

Reply:
Yeah, I would redirect the reader from this documentation to how `apply/dispatch + SignedExtensions` work in general to get more details into the lifecycle of extrinsic.

Reply:
Yeah, they are different checks in practice, but we are dealing with the non-pallete context, from the block builder API perspective, from that PoV they are essentially the same, isn't it This also relates to referencing the user to the documentation of `SignedExtensions` and co.